ubuntu_image/parser.py

```diff
@@ -320,41 +332,62 @@ def parse(stream_or_string):
             offset_write = structure.get('offset-write')
             size = structure['size']
             structure_type = structure['type']
-            # Validate structure types.  These can be either GUIDs, two hex
-            # digits, hybrids, or the special 'mbr' type.  The basic syntactic
-            # validation happens above in the Voluptuous schema, but here we
-            # need to ensure cross-attribute constraints.  Specifically,
-            # hybrids and 'mbr' are allowed for either schema, but GUID-only
-            # is only allowed for GPT, while 2-digit-only is only allowed for
-            # MBR.  Note too that 2-item tuples are also already ensured.
+            structure_role = structure.get('role')
+            is_mbr = \
+                structure_type == 'mbr' or structure_role is StructureRole.mbr
+            # Validate structure types and roles. These can be either GUIDs,
+            # two hex digits or hybrids.  The basic syntactic validation
+            # happens above in the Voluptuous schema, but here we need to
+            # ensure cross-attribute constraints.  Specifically, hybrids are
+            # allowed for either schema, but GUID-only is only allowed for GPT,
+            # while 2-digit-only is only allowed for MBR roles.
+            # Note too that 2-item tuples are also already ensured.
+            # We also still support the deprecated special type 'mbr' for
+            # legacy purposes, but issue a warning
+            if structure_type == 'mbr':
+                warn("volumes:<volume name>:structure:<N>:type = 'mbr' is "
+                     'deprecated; use role instead', DeprecationWarning)
+                if structure_role:
+                    raise GadgetSpecificationError(
+                        'Type mbr and role fields assigned at the same time, '
+                        'please use role field only')
```

Comment:
Imprecise error message, you don't check that structure_role is set to mbr here so 'please use role field only' could be directing them to do something that is syntactically correct but semantically not at all what they meant.